host/windows/syscall.c

```diff
@@ -366,34 +837,213 @@ oe_host_fd_t oe_syscall_open_ocall(
+        share_mode = FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE;
+        const int ACCESS_FLAGS = 0x3; // Covers rdonly, wronly rdwr
+        switch (flags & ACCESS_FLAGS)
+        {
+            case OE_O_RDONLY:
+            {
+                desired_access = GENERIC_READ;
+                break;
+            }
+            case OE_O_WRONLY:
+            {
+                desired_access =
+                    (flags & OE_O_APPEND) ? FILE_APPEND_DATA : GENERIC_WRITE;
+                break;
+            }
+            case OE_O_RDWR:
+            {
+                desired_access =
+                    GENERIC_READ |
+                    ((flags & OE_O_APPEND) ? FILE_APPEND_DATA : GENERIC_WRITE);
+                break;
+            }
+            default:
+                ret = -1;
+                _set_errno(OE_EINVAL);
+                goto done;
+                break;
+        }
+
+        if (mode & OE_S_IRUSR)
+            desired_access |= GENERIC_READ;
+        if (mode & OE_S_IWUSR)
+            desired_access |= GENERIC_WRITE;
+
+        HANDLE h = CreateFileW(
+            wpathname,
+            desired_access,
+            share_mode,
+            NULL,
+            create_dispos,
+            file_flags,
+            NULL);
+        if (h == INVALID_HANDLE_VALUE)
+        {
+            _set_errno(_winerr_to_errno(GetLastError()));
+            goto done;
+        }
+
+        ret = (oe_host_fd_t)h;
+
+        // Windows doesn't do mode in the same way as linux. We can set user
+        // read/write and thats about it. There are elaborate ACLs and the such
+        // for code which is purpose written, but the only part of file mode
+        // expressed in windows is the read-only bit, and only for the owner.
+        if (flags & OE_O_CREAT)
+        {
+            int wmode = ((mode & OE_S_IRUSR) ? _S_IREAD : 0) |
+                        ((mode & OE_S_IWUSR) ? _S_IWRITE : 0);
+
+            int retx = _wchmod(wpathname, wmode);
+            if (retx < 0)
+            {
+                _set_errno(_winerr_to_errno(GetLastError()));
+                goto done;
+            }
+        }
     }
 
 done:
-
+    if (wpathname)
+    {
+        free(wpathname);
+    }
     return ret;
 }
 
 ssize_t oe_syscall_read_ocall(oe_host_fd_t fd, void* buf, size_t count)
 {
-    if ((count & UINT_MAX) != count)
-        _set_errno(OE_EINVAL);
+    ssize_t ret = -1;
+    DWORD bytes_returned = 0;
 
-    return _read((int)fd, buf, (unsigned int)count);
+    HANDLE handle = (HANDLE)fd;
+
+    // Convert fd 0, 1, 2 as needed
+    switch (fd)
+    {
+        case OE_STDIN_FILENO:
+            handle = GetStdHandle(STD_INPUT_HANDLE);
+            break;
+
+        case OE_STDOUT_FILENO:
+        case OE_STDERR_FILENO:
+            _set_errno(OE_EBADF);
+            goto done;
+
+        default:
+            break;
+    }
+
+    if (!ReadFile(handle, buf, (DWORD)count, &bytes_returned, NULL))
```

Comment:
Again for `read()` and `write()` my concern is that Linux expects these to work for sockets as well as files, and we neither fail with a useful error, nor correctly redirect it to the socket implementation.

Reply:
Line 1387 uses a heuristic to determine whether a handle is a socket handle or a file handle, and the same check could be done here.

Reply:
Fixed with a helper function _is_valid_file_handle(), thanks.